packages/babel-generator/src/node/parentheses.js

```diff
@@ -166,20 +166,22 @@ export function UnaryLike(node: Object, parent: Object): boolean {
 }
 
 export function FunctionExpression(node: Object, parent: Object, printStack: Array<Object>): boolean {
-  return isFirstInStatement(printStack, { considerDefaultExports: true });
-}
-
-export function ArrowFunctionExpression(node: Object, parent: Object): boolean {
-  // export default (function () {});
-  if (t.isExportDeclaration(parent)) {
+  if (t.isTaggedTemplateExpression(parent)) {
```

Comment:
Oh wait, I see why you added this.

    function(){}`foo`

is fine as long as it is in an expression context, so I'd be in favor of expanding `isFirstInStatement` to handle functions in tagged templates, are you up for that? So your test output would be

    (() => {})``;
    (function(){}``);